src/workers/race_event_status_worker.rs

```diff
@@ -72,11 +89,156 @@ enum RaceEventContentAndStatus {
+
+        let mut shards = stream::create_recommended(&bot.client, cfg, |_, builder| builder.build())
+            .await
+            // TODO: surface this unwrap? idk
+            .unwrap()
+            .collect::<Vec<_>>();
+
+        // N.B. collecting these into a vec and then using `.iter_mut()` is stupid, but idk how to
+        // convince the compiler that i have an iterator of mutable references in a simpler way
+        let mut events = ShardEventStream::new(shards.iter_mut());
+
+        loop {
+            tokio::select! {
+                Some((_shard_id, evt)) = events.next() => {
+                    match evt {
+                        Ok(event) => {
+                            bot.cache.update(&event);
+                            bot.handle_event(event).await;
+                        }
+                        Err(e) => {
+                            warn!("Got error receiving discord event: {e}");
+                            if e.is_fatal() {
+                                info!("Helper bot shutting down due to fatal error");
+                                break;
+                            }
+                        }
+                    }
+                },
+
+                _sd = shutdown.recv() => {
+                    info!("Helper bot shutting down...");
+                    break;
+                }
+            }
+        }
+        info!("Helper bot done");
+    }
+
+    /// gets a [GuildEventConfig] for each guild we're currently in
+    ///
+    /// the CONFIG.guild_id guild will be first
+    fn guild_event_configs(&self) -> Vec<GuildEventConfig> {
+        let mut guild_ids = self
+            .cache
+            .iter()
+            .guilds()
+            .map(|guild| (guild.key().clone(), guild.value().name().to_string()))
+            .collect::<HashMap<_, _>>();
+        let mut out = Vec::with_capacity(guild_ids.len());
+
+        if let Some((guild_id, guild_name)) = guild_ids.remove_entry(&CONFIG.guild_id) {
+            out.push(GuildEventConfig {
+                guild_id,
+                guild_name,
+            });
+        }
+        out.extend(
+            guild_ids
+                .into_iter()
+                .map(|(guild_id, guild_name)| GuildEventConfig {
+                    guild_id,
+                    guild_name,
+                }),
+        );
+        out
+    }
+
+    async fn handle_event(&self, event: Event) {
+        match event {
+            Event::GuildCreate(gc) => {
+                info!("Joined guild {}: {}", gc.id, gc.name);
+            }
+            _ => {}
+        }
+    }
+}
+
+pub async fn launch(
+    mut sd: Receiver<Shutdown>,
+    state: Arc<DiscordState>,
+    pool: Pool<DieselConnectionManager>,
+) {
     let mut intv = tokio::time::interval(Duration::from_secs(CONFIG.race_event_worker_tick_secs));
+    let bot = Arc::new(HelperBot::new(pool));
+    tokio::spawn(HelperBot::run(bot.clone(), sd.resubscribe()));
```

Comment:
every time i write the pattern of a "a struct sharing state across threads" i find a slightly different way to do it

this is probably not good, overall, but i do like this way i think...